<h1><center>Final Project - Omni Oracle</center></h1>
<hr>

<table style="margin-left: auto; margin-right: auto;">
  <tr><th>Team members</th></tr>
  <tr><td>Francis Tan Hong Xin</td></tr>
  <tr><td>Gan Lai Soon</td></tr>
  <tr><td>Khadijah Anhardeen</td></tr>
  <tr><td>Muhammad Nur Syafaat Bin Mohamed Saat</td></tr>
</table>

<h2><center>Table of contents<center></h2>
<hr>

In [27]:
!pip install unidecode
!pip install pyspellchecker

   ---------------------------------------- 0.0/6.8 MB ? eta -:--:--
   ---------------------------------------- 0.0/6.8 MB ? eta -:--:--
   ---------------------------------------- 0.0/6.8 MB 435.7 kB/s eta 0:00:16
    --------------------------------------- 0.1/6.8 MB 1.1 MB/s eta 0:00:07
   ---- ----------------------------------- 0.7/6.8 MB 4.9 MB/s eta 0:00:02
   ------- -------------------------------- 1.4/6.8 MB 7.1 MB/s eta 0:00:01
   ------------- -------------------------- 2.2/6.8 MB 9.3 MB/s eta 0:00:01
   ----------------- ---------------------- 2.9/6.8 MB 10.3 MB/s eta 0:00:01
   --------------------- ------------------ 3.7/6.8 MB 11.2 MB/s eta 0:00:01
   ------------------------- -------------- 4.4/6.8 MB 11.6 MB/s eta 0:00:01
   ------------------------------- -------- 5.3/6.8 MB 12.5 MB/s eta 0:00:01
   ---------------------------------- ----- 5.8/6.8 MB 12.8 MB/s eta 0:00:01
   ------------------------------------- -- 6.4/6.8 MB 12.4 MB/s eta 0:00:01
   ---------------

In [1]:
# for DataFrames, Series, read_csv(), to_csv(), etc.
import pandas as pd

# for math, data types
import numpy as np

# for dates, times, etc.
import datetime as dt

# for recording logs
import logging

# for foreign characters (Portuguese) and converting them
from unidecode import unidecode

from spellchecker import SpellChecker

# for connecting to PostgreSQL/pgAdmin and inserting data to database
import sqlalchemy as db
from sqlalchemy_utils import create_database
from sqlalchemy_utils import database_exists
from sqlalchemy.types import Integer, String, Text, ARRAY, VARCHAR, Date, Numeric, BigInteger

In [2]:
# Configure logging
logging.basicConfig(level=logging.INFO, filename='etl_pipeline.log', 
                    format='%(asctime)s:%(levelname)s:%(message)s')

logging.info("ETL process started.")

# Extraction

In [3]:
# Reading geolocation data
geolocations_data_file_path = '../resources/olist_geolocation_dataset.csv'
geolocations_col_names = [
    'zip_code_prefix', 
    'latitude', 
    'longitude', 
    'city', 
    'state'
]
geolocations_data_types = {
    "zip_code_prefix": np.int32,
    "latitude": np.float64,
    "longitude": np.float64,
    "city": "string",
    "state": "string"
}

geolocations_df = pd.read_csv(geolocations_data_file_path,
                              header=0,
                              names=geolocations_col_names,
                              dtype=geolocations_data_types,
                              na_values='')

logging.info("Geolocation data loaded successfully.")

geolocations_na = geolocations_df.isna().any().to_dict()
if True in geolocations_na.values():
    logging.warning("Detected columns with NA values: ", [col for col in geolocations_na.keys() if geolocations_na[col] == True])

geolocations_df.head(10)

zip_code_prefix   latitude  longitude       city state
0             1037 -23.545621 -46.639292  sao paulo    SP
1             1046 -23.546081 -46.644820  sao paulo    SP
2             1046 -23.546129 -46.642951  sao paulo    SP
3             1041 -23.544392 -46.639499  sao paulo    SP
4             1035 -23.541578 -46.641607  sao paulo    SP
5             1012 -23.547762 -46.635361  são paulo    SP
6             1047 -23.546273 -46.641225  sao paulo    SP
7             1013 -23.546923 -46.634264  sao paulo    SP
8             1029 -23.543769 -46.634278  sao paulo    SP
9             1011 -23.547640 -46.636032  sao paulo    SP

In [4]:
# Reading customers data
customers_data_file_path = '../resources/olist_customers_dataset.csv'
customers_data_types = {
    "customer_id": "string",
    "customer_unique_id": "string",
    "customer_zip_code_prefix": np.int32
}

customers_df = pd.read_csv(customers_data_file_path,
                           header=0,
                           usecols=[0,1,2],
                           dtype=customers_data_types,
                           na_values=np.nan)

logging.info("Customer data loaded successfully.")

customers_df.head(10)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
5  879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4   
6  fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177   
7  5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f   
8  5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d   
9  4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22   

   customer_zip_code_prefix  
0                     14409  
1                      9790  
2                      1151  
3                      8775  
4                     13056  
5                     89254  
6                      4534  
7                     35182  
8                     81560  
9                     30575

In [5]:
sellers_data_file_path = '../resources/olist_sellers_dataset.csv'
sellers_data_types = {
    "seller_id": "string",
    "seller_zip_code_prefix": np.int32
}

sellers_df = pd.read_csv(sellers_data_file_path,
                        header=0,
                        usecols=[0,1],
                        dtype=sellers_data_types)

logging.info("Seller data loaded successfully.")

sellers_df.head(10)

seller_id  seller_zip_code_prefix
0  3442f8959a84dea7ee197c632cb2df15                   13023
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914
5  c240c4061717ac1806ae6ee72be3533b                   20920
6  e49c26c3edfa46d227d5121a6b6e4d37                   55325
7  1b938a7ec6ac5061a66a3766e0e75f90                   16304
8  768a86e36ad6aae3d03ee3c6433d61df                    1529
9  ccc4bbb5f32a6ab2b7066a4130f114e3                   80310

In [6]:
orders_data_file_path = '../resources/olist_orders_dataset.csv'
orders_data_types = {
    "order_id": "string",
    "customer_id": "string",
    "order_status": "string"
}
orders_date_cols = [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

orders_df = pd.read_csv(orders_data_file_path,
                       header=0,
                       dtype=orders_data_types,
                       parse_dates=orders_date_cols)

logging.info("Order data loaded successfully.")

orders_df.head(10)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
5  a4591c265e18cb1dcee52889e2d8acc3  503740e9ca751ccdda7ba28e9ab8f608   
6  136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
7  6514b8ad8028c9f2cc2374ded245783f  9bdf08b4b3b52b5526ff42d37d47f222   
8  76c6e866289321a7c93b82b54852dc33  f54a9f0e6b351c431402b8461ea51999   
9  e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   
5    delivered      2017-07-09 21:57:05 2017-07-09 22:10:13   
6     invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
7    delivered      2017-05-16 13:10:30 2017-05-16 13:22:11   
8    delivered      2017-01-23 18:29:09 2017-01-25 02:50:47   
9    delivered      2017-07-29 11:55:02 2017-07-29 12:05:32   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   
5          2017-07-11 14:58:04           2017-07-26 10:57:55   
6                          NaT                           NaT   
7          2017-05-22 10:07:46           2017-05-26 12:55:51   
8          2017-01-26 14:16:31           2017-02-02 14:08:10   
9          2017-08-10 19:45:24           2017-08-16 17:14:30   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26  
5                    2017-08-01  
6                    2017-05-09  
7                    2017-06-07  
8                    2017-03-06  
9                    2017-08-23

In [7]:
product_categories_data_file_path = '../resources/product_category_name_translation.csv'
product_categories_data_types = {
    "product_category_name": "string",
    "product_category_name_english": "string"
}

product_categories_df = pd.read_csv(product_categories_data_file_path,
                                    header=0,
                                    dtype=product_categories_data_types)

logging.info("Product category data loaded successfully.")

product_categories_df.head(10)

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor
5           esporte_lazer                sports_leisure
6              perfumaria                     perfumery
7   utilidades_domesticas                    housewares
8               telefonia                     telephony
9      relogios_presentes                 watches_gifts

In [8]:
products_data_file_path = '../resources/olist_products_dataset.csv'
products_col_names = [
    "product_id",
    "product_category_name",
    "product_name_length",
    "product_description_length",
    "product_photos_qty",
    "product_weight_g",
    "product_length_cm",
    "product_height_cm",
    "product_width_cm"
]
products_data_types = {
    "product_id": "string",
    "product_category_name": "string",
    "product_name_length": np.int8,
    "product_description_length": np.int16,
    "product_photos_qty": np.int8,
    "product_weight_g": np.int16,
    "product_length_cm": np.int8,
    "product_height_cm": np.int8,
    "product_width_cm": np.int8
}
products_na_values = {
    "product_name_length": 0,
    "product_description_length": 0,
    "product_photos_qty": 0,
    "product_weight_g": 0,
    "product_length_cm": 0,
    "product_height_cm": 0,
    "product_width_cm": 0
}

products_df = pd.read_csv(products_data_file_path,
                         header=0,
                         names=products_col_names,
                         na_values='')

products_df.dropna(subset=['product_category_name'], inplace=True)
products_df.fillna(value=products_na_values, inplace=True)

products_df = products_df.astype(products_data_types)

logging.info("Product data loaded successfully.")

products_df.head(10)

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   
5  41d3672d4792049fa1779bb35283ed13  instrumentos_musicais   
6  732bd381ad09e530fe0a5f457d81becb             cool_stuff   
7  2548af3e6e77a690cf3eb6368e9ab61e       moveis_decoracao   
8  37cc742be07708b53a98702e77a21a02       eletrodomesticos   
9  8c92109888e8cdf9d66dc7e463025574             brinquedos   

   product_name_length  product_description_length  product_photos_qty  \
0                   40                         287                   1   
1                   44                         276                   1   
2                   46                         250                   1   
3                   27                         261                   1   
4                   37                         402                   4   
5                   60                         745                   1   
6                   56                        1272                   4   
7                   56                         184                   2   
8                   57                         163                   1   
9                   36                        1156                   1   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0               225                 16                 10                14  
1              1000                 30                 18                20  
2               154                 18                  9                15  
3               371                 26                  4                26  
4               625                 20                 17                13  
5               200                 38                  5                11  
6             18350                 70                 24                44  
7               900                 40                  8                40  
8               400                 27                 13                17  
9               600                 17                 10                12

In [9]:
order_reviews_data_file_path = '../resources/olist_order_reviews_dataset.csv'
order_reviews_data_types = {
    "review_id": "string",
    "order_id": "string",
    "review_score": np.int8,
    "review_comment_title": "string",
    "review_comment_message": "string"
}
order_reviews_date_cols = [
    "review_creation_date",
    "review_answer_timestamp"
]

order_reviews_df = pd.read_csv(order_reviews_data_file_path,
                               header=0,
                               parse_dates=order_reviews_date_cols,
                               dtype=order_reviews_data_types)

logging.info("Order review data loaded successfully.")

order_reviews_df.head(10)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
5  15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
6  07f9bee5d1b850860defd761afa7ff16  e48aa0d2dcec3a2e87348811bcfdf22b   
7  7c6400515c67679fbee952a7525281ef  c31a859e34e3adac22f376954e19b39d   
8  a3f6f7f6f433de0aefbb97da197c554c  9c214ac970e84273583ab523dfafd09b   
9  8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   

   review_score review_comment_title  \
0             4                 <NA>   
1             5                 <NA>   
2             5                 <NA>   
3             5                 <NA>   
4             5                 <NA>   
5             1                 <NA>   
6             5                 <NA>   
7             5                 <NA>   
8             5                 <NA>   
9             4            recomendo   

                              review_comment_message review_creation_date  \
0                                               <NA>           2018-01-18   
1                                               <NA>           2018-03-10   
2                                               <NA>           2018-02-17   
3              Recebi bem antes do prazo estipulado.           2017-04-21   
4  Parabéns lojas lannister adorei comprar pela I...           2018-03-01   
5                                               <NA>           2018-04-13   
6                                               <NA>           2017-07-16   
7                                               <NA>           2018-08-14   
8                                               <NA>           2017-05-17   
9  aparelho eficiente. no site a marca do aparelh...           2018-05-22   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53  
5     2018-04-16 00:39:37  
6     2017-07-18 19:30:34  
7     2018-08-14 21:36:06  
8     2017-05-18 12:05:37  
9     2018-05-23 16:45:47

In [10]:
order_items_data_file_path = '../resources/olist_order_items_dataset.csv'
order_items_data_types = {
    "order_id": "string",
    "order_item_id": np.int8,
    "product_id": "string",
    "seller_id": "string",
    "price": np.float64,
    "freight_value": np.float64
}
order_items_date_cols = [
    "shipping_limit_date"
]

order_items_df = pd.read_csv(order_items_data_file_path,
                             header=0,
                             parse_dates=order_items_date_cols,
                             dtype=order_items_data_types,
                             na_values='')

logging.info("Order item data loaded successfully.")

order_items_df.head(10)

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   
5  00048cc3ae777c65dbb7d2a0634bc1ea              1   
6  00054e8431b9d7675808bcb819fb4a32              1   
7  000576fe39319847cbb9d288c5617fa6              1   
8  0005a1a1728c9d785b8e2b08b904576c              1   
9  0005f50442cb953dcd1d21e1fb923495              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
5  ef92defde845ab8450f9d70c526ef70f  6426d21aca402a131fc0a5d0960a3c90   
6  8d4f2bb7e93e6710a28f34fa83ee7d28  7040e82f899a04d1b434b795a43b4617   
7  557d850972a7d6f792fd18ae1400d9b6  5996cddab893a4652a15592fb58ab8db   
8  310ae3c140ff94b03219ad0adc3c778f  a416b6a846a11724393025641d4edd5e   
9  4535b0e1091c278dfd193e5a1d63b39f  ba143b05f0110f0dc71ad71b4466ce92   

  shipping_limit_date   price  freight_value  
0 2017-09-19 09:45:35   58.90          13.29  
1 2017-05-03 11:05:13  239.90          19.93  
2 2018-01-18 14:48:30  199.00          17.87  
3 2018-08-15 10:10:18   12.99          12.79  
4 2017-02-13 13:57:51  199.90          18.14  
5 2017-05-23 03:55:27   21.90          12.69  
6 2017-12-14 12:10:31   19.90          11.85  
7 2018-07-10 12:30:45  810.00          70.75  
8 2018-03-26 18:31:29  145.95          11.65  
9 2018-07-06 14:10:56   53.99          11.40

In [11]:
order_payments_data_file_path = '../resources/olist_order_payments_dataset.csv'
order_payments_data_types = {
    "order_id": "string",
    "payment_sequential": np.int8,
    "payment_type": "string",
    "payment_installments": np.int8,
    "payment_value": np.float64
}

order_payments_df = pd.read_csv(order_payments_data_file_path,
                                header=0,
                                dtype=order_payments_data_types,
                                na_values='')

logging.info("Order payment data loaded successfully.")

order_payments_df.head(10)

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
5  298fcdf1f73eb413e4d26d01b25bc1cd                   1  credit_card   
6  771ee386b001f06208a7419e4fc1bbd7                   1  credit_card   
7  3d7239c394a212faae122962df514ac7                   1  credit_card   
8  1f78449c87a54faf9e96e88ba1491fa9                   1  credit_card   
9  0573b5e23cbd798006520e1d5b4c6714                   1       boleto   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45  
5                     2          96.12  
6                     1          81.16  
7                     3          51.84  
8                     6         341.09  
9                     1          51.95

In [12]:
logging.info("Extraction process completed.")

# Transformation

In [13]:
def get_mode(series):
    mode = series.mode()
    if not mode.empty:
        return mode.iloc[0]
    else:
        return None

In [14]:
spell = SpellChecker()
spell.word_frequency.load_text_file('../resources/product_categories.txt')

def correct_spelling(text):
    list_of_words = text.split()
    for i in range(len(list_of_words)):
        word = list_of_words[i]
        list_of_words[i] = spell.correction(word)
    return ' '.join(list_of_words)

<hr>
<h3>Geolocations dataset</h3>

#### Problems:

1. The strings under the `city` column has different characters with some using foreign (Portuguese) characters like 'ã', 'ç', 'é', etc.

2. There are duplicate `zip_code_prefix` values which makes it a problem as it is supposed to be the primary key of this table.

3. Because we want to merge the duplicate `zip_code_prefix` values into one, another problem arise where there are multiple different `city` and `state` values under one unique `zip_code_prefix`.

In [15]:
geolocations_df.sort_values(by=['zip_code_prefix', 'city', 'state'], ascending=[True, False, False]).head(10)

zip_code_prefix   latitude  longitude       city state
575              1001 -23.549779 -46.633957  são paulo    SP
608              1001 -23.550263 -46.634196  são paulo    SP
1351             1001 -23.549951 -46.634027  são paulo    SP
99               1001 -23.549292 -46.633559  sao paulo    SP
206              1001 -23.550498 -46.634338  sao paulo    SP
235              1001 -23.550642 -46.634410  sao paulo    SP
299              1001 -23.549698 -46.633909  sao paulo    SP
326              1001 -23.551427 -46.634074  sao paulo    SP
429              1001 -23.550498 -46.634338  sao paulo    SP
519              1001 -23.551337 -46.634027  sao paulo    SP

In [16]:
before_transform_shape = geolocations_df.shape
before_transform_shape

(1000163, 5)

#### Change the data in 'city' to lowercase and the data in 'state' to uppercase.

This makes it easier for us to handle the conversion and grouping process later on.

In [17]:
geolocations_df['city'] = geolocations_df['city'].str.lower()
geolocations_df['state'] = geolocations_df['state'].str.upper()

geolocations_df.head(5)

zip_code_prefix   latitude  longitude       city state
0             1037 -23.545621 -46.639292  sao paulo    SP
1             1046 -23.546081 -46.644820  sao paulo    SP
2             1046 -23.546129 -46.642951  sao paulo    SP
3             1041 -23.544392 -46.639499  sao paulo    SP
4             1035 -23.541578 -46.641607  sao paulo    SP

#### Normalize the data in `city` to use normal characters, replacing any foreign characters like 'ã', 'ç', 'é', etc. 

This is so that it makes it easier to group the data according to the `zip_code_prefix`. 

We have also noticed that for some `zip_code_prefix`, there are different cities and states under some of them. To set the `city` and `state` for one `zip_code_prefix`, we choose the `city` and `state` that has the highest occurence within a `zip_code_prefix`. 

We will be using the **unidecode** library to detect the foreign characters and convert them to their normal character counterparts. We store the converted `city` data into another column called `city_normalized`.

As for the `latitude` and `longitude`, we will be taking their average values grouped by the `zip_code_prefix`.

We will assign the aggregated data into a temporary DataFrame called `aggregated_geo_df`.

In [18]:
geolocations_df['city_normalized'] = geolocations_df['city'].apply(unidecode)
statistics = {
                'city_normalized': get_mode,
                'state': get_mode,
                'latitude': 'mean',
                'longitude': 'mean'
             }

aggregated_geo_df = geolocations_df.groupby('zip_code_prefix').agg(statistics).reset_index()
aggregated_geo_df.head(10)

zip_code_prefix city_normalized state   latitude  longitude
0             1001       sao paulo    SP -23.550190 -46.634024
1             1002       sao paulo    SP -23.548146 -46.634979
2             1003       sao paulo    SP -23.548994 -46.635731
3             1004       sao paulo    SP -23.549799 -46.634757
4             1005       sao paulo    SP -23.549456 -46.636733
5             1006       sao paulo    SP -23.550102 -46.636137
6             1007       sao paulo    SP -23.550046 -46.637251
7             1008       sao paulo    SP -23.546002 -46.635886
8             1009       sao paulo    SP -23.546835 -46.636491
9             1010       sao paulo    SP -23.546389 -46.635226

#### Rename columns and assigning data to original DataFrame.

We need to rename the `city_normalized` column back to `city` and assign the aggregated DataFrame back to the original `geolocations_df` DataFrame. We will also export the dataset to a CSV file for reference.

In [19]:
aggregated_geo_df.rename(columns={'city_normalized': 'city'}, inplace=True)
geolocations_df = aggregated_geo_df[['zip_code_prefix', 'latitude', 'longitude', 'city', 'state']]

cleaned_geolocations_data_file_path = '../resources/cleaned_geolocation_dataset.csv'
geolocations_df.to_csv(cleaned_geolocations_data_file_path, index=False)

logging.info(f"Cleaned geolocation data and exported to {cleaned_geolocations_data_file_path}.")

geolocations_df.head(10)

zip_code_prefix   latitude  longitude       city state
0             1001 -23.550190 -46.634024  sao paulo    SP
1             1002 -23.548146 -46.634979  sao paulo    SP
2             1003 -23.548994 -46.635731  sao paulo    SP
3             1004 -23.549799 -46.634757  sao paulo    SP
4             1005 -23.549456 -46.636733  sao paulo    SP
5             1006 -23.550102 -46.636137  sao paulo    SP
6             1007 -23.550046 -46.637251  sao paulo    SP
7             1008 -23.546002 -46.635886  sao paulo    SP
8             1009 -23.546835 -46.636491  sao paulo    SP
9             1010 -23.546389 -46.635226  sao paulo    SP

#### Difference before and after cleaning:

As you can see below, the **geolocations_df** DataFrame lost **981,148** records due to duplicate `zip_code_prefix` values from **1,000,163** records to **19,015** records.

In [20]:
before_transform_shape

(1000163, 5)

In [21]:
geolocations_df.shape

(19015, 5)

<hr>
<h3>Customers dataset</h3>

#### Problems:

1. We need to check if the data under the `customer_zip_code_prefix` column do exist inside the **geolocations_df** DataFrame.

In [22]:
before_transform_shape = customers_df.shape
before_transform_shape

(99441, 3)

#### Remove rows where the `customer_zip_code_prefix` does not exist inside the **geolocations_df** DataFrame.

In [23]:
customers_df = customers_df[customers_df['customer_zip_code_prefix'].isin(geolocations_df['zip_code_prefix'])]

cleaned_customers_data_file_path = '../resources/cleaned_customer_dataset.csv'
customers_df.to_csv(cleaned_customers_data_file_path, index=False)

logging.info(f"Cleaned customer data and exported to {cleaned_customers_data_file_path}.")

customers_df.head(10)

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   
5  879864dab9bc3047522c92c82e1212b8  4c93744516667ad3b8f1fb645a3116a4   
6  fd826e7cf63160e536e0908c76c3f441  addec96d2e059c80c30fe6871d30d177   
7  5e274e7a0c3809e14aba7ad5aae0d407  57b2a98a409812fe9618067b6b8ebe4f   
8  5adf08e34b2e993982a47070956c5c65  1175e95fb47ddff9de6b2b06188f7e0d   
9  4b7139f34592b3a31687243a302fa75b  9afe194fb833f79e300e37e580171f22   

   customer_zip_code_prefix  
0                     14409  
1                      9790  
2                      1151  
3                      8775  
4                     13056  
5                     89254  
6                      4534  
7                     35182  
8                     81560  
9                     30575

#### Difference before and after cleaning:

As you can see below, the **customers_df** DataFrame lost **278** records due to missing `zip_code_prefix` values from **99,441** records to **99,163** records.

In [24]:
before_transform_shape

(99441, 3)

In [25]:
customers_df.shape

(99163, 3)

<hr>
<h3>Sellers dataset</h3>

In [26]:
sellers_df = sellers_df[sellers_df['seller_zip_code_prefix'].isin(geolocations_df['zip_code_prefix'])]

cleaned_sellers_data_file_path = '../resources/cleaned_seller_dataset.csv'
sellers_df.to_csv(cleaned_sellers_data_file_path, index=False)

logging.info(f"Cleaned seller data and exported to {cleaned_sellers_data_file_path}.")

sellers_df.head(10)

seller_id  seller_zip_code_prefix
0  3442f8959a84dea7ee197c632cb2df15                   13023
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914
5  c240c4061717ac1806ae6ee72be3533b                   20920
6  e49c26c3edfa46d227d5121a6b6e4d37                   55325
7  1b938a7ec6ac5061a66a3766e0e75f90                   16304
8  768a86e36ad6aae3d03ee3c6433d61df                    1529
9  ccc4bbb5f32a6ab2b7066a4130f114e3                   80310

In [27]:
orders_df = orders_df[orders_df['customer_id'].isin(customers_df['customer_id'])]

today_date = dt.date.today()
orders_df.drop(orders_df[orders_df['order_purchase_timestamp'].dt.date >= today_date].index, inplace=True)
orders_df.drop(orders_df[orders_df['order_approved_at'].dt.date >= today_date].index, inplace=True)
orders_df.drop(orders_df[orders_df['order_delivered_carrier_date'].dt.date >= today_date].index, inplace=True)
orders_df.drop(orders_df[orders_df['order_delivered_customer_date'].dt.date >= today_date].index, inplace=True)
orders_df.drop(orders_df[orders_df['order_estimated_delivery_date'].dt.date >= today_date].index, inplace=True)

cleaned_orders_data_file_path = '../resources/cleaned_order_dataset.csv'
orders_df.to_csv(cleaned_orders_data_file_path, index=False)

logging.info(f"Cleaned order data and exported to {cleaned_orders_data_file_path}.")


In [28]:
product_categories_df.head(10)

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor
5           esporte_lazer                sports_leisure
6              perfumaria                     perfumery
7   utilidades_domesticas                    housewares
8               telefonia                     telephony
9      relogios_presentes                 watches_gifts

In [29]:
product_categories_df['product_category_name'] = product_categories_df['product_category_name'].str.replace('_', ' ')
product_categories_df['product_category_name_english'] = product_categories_df['product_category_name_english'].str.replace('_', ' ')
product_categories_df['product_category_name_english'] = product_categories_df['product_category_name_english'].apply(correct_spelling)

cleaned_products_categories_data_file_path = '../resources/cleaned_product_categories_dataset.csv'
product_categories_df.to_csv(cleaned_products_categories_data_file_path, index=False)

logging.info(f"Cleaned product data and exported to {cleaned_products_categories_data_file_path}.")

product_categories_df.head(10)

product_category_name product_category_name_english
0            beleza saude                 health beauty
1  informatica acessorios         computers accessories
2              automotivo                          auto
3         cama mesa banho                bed bath table
4        moveis decoracao               furniture decor
5           esporte lazer                sports leisure
6              perfumaria                     perfumery
7   utilidades domesticas                    housewares
8               telefonia                     telephony
9      relogios presentes                 watches gifts

In [30]:
products_df['product_category_name'] = products_df['product_category_name'].str.replace('_', ' ')
products_df = products_df[products_df['product_category_name'].isin(product_categories_df['product_category_name'])]

cleaned_products_data_file_path = '../resources/cleaned_product_dataset.csv'
products_df.to_csv(cleaned_products_data_file_path, index=False)

logging.info(f"Cleaned product data and exported to {cleaned_products_data_file_path}.")

products_df.head(10)

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades domesticas   
5  41d3672d4792049fa1779bb35283ed13  instrumentos musicais   
6  732bd381ad09e530fe0a5f457d81becb             cool stuff   
7  2548af3e6e77a690cf3eb6368e9ab61e       moveis decoracao   
8  37cc742be07708b53a98702e77a21a02       eletrodomesticos   
9  8c92109888e8cdf9d66dc7e463025574             brinquedos   

   product_name_length  product_description_length  product_photos_qty  \
0                   40                         287                   1   
1                   44                         276                   1   
2                   46                         250                   1   
3                   27                         261                   1   
4                   37                         402                   4   
5                   60                         745                   1   
6                   56                        1272                   4   
7                   56                         184                   2   
8                   57                         163                   1   
9                   36                        1156                   1   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0               225                 16                 10                14  
1              1000                 30                 18                20  
2               154                 18                  9                15  
3               371                 26                  4                26  
4               625                 20                 17                13  
5               200                 38                  5                11  
6             18350                 70                 24                44  
7               900                 40                  8                40  
8               400                 27                 13                17  
9               600                 17                 10                12

In [31]:
order_reviews_df = order_reviews_df[order_reviews_df['order_id'].isin(orders_df['order_id'])]

cleaned_order_reviews_data_file_path = '../resources/cleaned_order_reviews_dataset.csv'
order_reviews_df.to_csv(cleaned_order_reviews_data_file_path, index=False)

logging.info(f"Cleaned order review data and exported to {cleaned_order_reviews_data_file_path}.")

order_reviews_df.head(10)

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
5  15197aa66ff4d0650b5434f1b46cda19  b18dcdf73be66366873cd26c5724d1dc   
6  07f9bee5d1b850860defd761afa7ff16  e48aa0d2dcec3a2e87348811bcfdf22b   
7  7c6400515c67679fbee952a7525281ef  c31a859e34e3adac22f376954e19b39d   
8  a3f6f7f6f433de0aefbb97da197c554c  9c214ac970e84273583ab523dfafd09b   
9  8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   

   review_score review_comment_title  \
0             4                 <NA>   
1             5                 <NA>   
2             5                 <NA>   
3             5                 <NA>   
4             5                 <NA>   
5             1                 <NA>   
6             5                 <NA>   
7             5                 <NA>   
8             5                 <NA>   
9             4            recomendo   

                              review_comment_message review_creation_date  \
0                                               <NA>           2018-01-18   
1                                               <NA>           2018-03-10   
2                                               <NA>           2018-02-17   
3              Recebi bem antes do prazo estipulado.           2017-04-21   
4  Parabéns lojas lannister adorei comprar pela I...           2018-03-01   
5                                               <NA>           2018-04-13   
6                                               <NA>           2017-07-16   
7                                               <NA>           2018-08-14   
8                                               <NA>           2017-05-17   
9  aparelho eficiente. no site a marca do aparelh...           2018-05-22   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53  
5     2018-04-16 00:39:37  
6     2017-07-18 19:30:34  
7     2018-08-14 21:36:06  
8     2017-05-18 12:05:37  
9     2018-05-23 16:45:47

In [32]:
order_items_df = order_items_df[order_items_df['order_id'].isin(orders_df['order_id'])]
order_items_df = order_items_df[order_items_df['product_id'].isin(products_df['product_id'])]
order_items_df = order_items_df[order_items_df['seller_id'].isin(sellers_df['seller_id'])]

cleaned_order_items_data_file_path = '../resources/cleaned_order_items_dataset.csv'
order_items_df.to_csv(cleaned_order_items_data_file_path, index=False)

logging.info(f"Cleaned order item data and exported to {cleaned_order_items_data_file_path}.")

order_items_df.head(10)

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   
5  00048cc3ae777c65dbb7d2a0634bc1ea              1   
6  00054e8431b9d7675808bcb819fb4a32              1   
7  000576fe39319847cbb9d288c5617fa6              1   
8  0005a1a1728c9d785b8e2b08b904576c              1   
9  0005f50442cb953dcd1d21e1fb923495              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   
5  ef92defde845ab8450f9d70c526ef70f  6426d21aca402a131fc0a5d0960a3c90   
6  8d4f2bb7e93e6710a28f34fa83ee7d28  7040e82f899a04d1b434b795a43b4617   
7  557d850972a7d6f792fd18ae1400d9b6  5996cddab893a4652a15592fb58ab8db   
8  310ae3c140ff94b03219ad0adc3c778f  a416b6a846a11724393025641d4edd5e   
9  4535b0e1091c278dfd193e5a1d63b39f  ba143b05f0110f0dc71ad71b4466ce92   

  shipping_limit_date   price  freight_value  
0 2017-09-19 09:45:35   58.90          13.29  
1 2017-05-03 11:05:13  239.90          19.93  
2 2018-01-18 14:48:30  199.00          17.87  
3 2018-08-15 10:10:18   12.99          12.79  
4 2017-02-13 13:57:51  199.90          18.14  
5 2017-05-23 03:55:27   21.90          12.69  
6 2017-12-14 12:10:31   19.90          11.85  
7 2018-07-10 12:30:45  810.00          70.75  
8 2018-03-26 18:31:29  145.95          11.65  
9 2018-07-06 14:10:56   53.99          11.40

In [33]:
order_payments_df = order_payments_df[order_payments_df['order_id'].isin(orders_df['order_id'])]

cleaned_order_payments_data_file_path = '../resources/cleaned_order_payments_dataset.csv'
order_payments_df.to_csv(cleaned_order_payments_data_file_path, index=False)

logging.info(f"Cleaned order payment data and exported to {cleaned_order_payments_data_file_path}.")

order_payments_df.head(10)

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   
5  298fcdf1f73eb413e4d26d01b25bc1cd                   1  credit_card   
6  771ee386b001f06208a7419e4fc1bbd7                   1  credit_card   
7  3d7239c394a212faae122962df514ac7                   1  credit_card   
8  1f78449c87a54faf9e96e88ba1491fa9                   1  credit_card   
9  0573b5e23cbd798006520e1d5b4c6714                   1       boleto   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45  
5                     2          96.12  
6                     1          81.16  
7                     3          51.84  
8                     6         341.09  
9                     1          51.95

In [34]:
logging.info("Transformation process completed.")

# Load

In [35]:
database_name = 'omni_oracle_ecommerce_db'

# user postgres, password admin
db_engine = db.create_engine(f'postgresql://postgres:admin@localhost:5432/{database_name}')
if database_exists(f'postgresql://postgres:admin@localhost:5432/{database_name}'):
    logging.info(f"{database_name} already exists.")
else:
    # create database
    create_database(db_engine.url)
    logging.info(f"Created database: {database_name}")

conn = db_engine.raw_connection()

# release resources associated with engine
db_engine.dispose()

# Initialize connection to PostgreSQL
cur = conn.cursor()

In [36]:
# Create new tables in PostgreSQL
commands = ('''
            -- geolocations
            CREATE TABLE geolocations (
                zip_code_prefix INTEGER PRIMARY KEY,
                latitude NUMERIC,
                longitude NUMERIC,
                city VARCHAR(50),
                state VARCHAR(10)
            );

            -- customers
            CREATE TABLE customers (
                customer_id VARCHAR(40) PRIMARY KEY,
                customer_unique_id VARCHAR(40),
                customer_zip_code_prefix INTEGER REFERENCES geolocations(zip_code_prefix)
            );

            -- sellers
            CREATE TABLE sellers (
                seller_id VARCHAR(40) PRIMARY KEY,
                seller_zip_code_prefix INTEGER REFERENCES geolocations(zip_code_prefix)
            );

            -- orders
            CREATE TABLE orders (
                order_id VARCHAR(40) PRIMARY KEY,
                customer_id VARCHAR(40) REFERENCES customers (customer_id),
                order_status VARCHAR(20),
                order_purchase_timestamp TIMESTAMP,
                order_approved_at TIMESTAMP,
                order_delivered_carrier_date TIMESTAMP,
                order_delivered_customer_date TIMESTAMP,
                order_estimated_delivery_date TIMESTAMP
            );

            -- product_category_name_translations
            CREATE TABLE product_category_name_translations (
                product_category_name TEXT PRIMARY KEY,
                product_category_name_english TEXT
            );

            -- products
            CREATE TABLE products (
                product_id VARCHAR(40) PRIMARY KEY,
                product_category_name TEXT REFERENCES product_category_name_translations (product_category_name),
                product_name_length INTEGER,
                product_description_length INTEGER,
                product_photos_qty INTEGER,
                product_weight_g INTEGER,
                product_length_cm INTEGER,
                product_height_cm INTEGER,
                product_width_cm INTEGER
            );

            -- order_reviews
            CREATE TABLE order_reviews (
                review_id VARCHAR(40),
                order_id VARCHAR(40) REFERENCES orders (order_id),
                review_score INTEGER,
                review_comment_title TEXT,
                review_comment_message TEXT,
                review_creation_date TIMESTAMP,
                review_answer_timestamp TIMESTAMP,
                PRIMARY KEY(review_id, order_id)
            );

            -- order_items
            CREATE TABLE order_items (
                order_id VARCHAR(40) REFERENCES orders (order_id),
                order_item_id INTEGER,
                product_id VARCHAR(40) REFERENCES products (product_id),
                seller_id VARCHAR(40) REFERENCES sellers (seller_id),
                shipping_limit_date TIMESTAMP,
                price MONEY,
                freight_value MONEY,
                PRIMARY KEY(order_id, order_item_id, product_id, seller_id)
            );

            -- order_payments
            CREATE TABLE order_payments (
                order_id VARCHAR(40) REFERENCES orders (order_id),
                payment_sequential INTEGER,
                payment_type VARCHAR(20),
                payment_installments INTEGER,
                payment_value MONEY,
                PRIMARY KEY(order_id, payment_sequential)
            );
            ''')

# Create cursor to execute SQL commands
cur.execute(commands)

# Commit changes
conn.commit()

logging.info("Created tables in database.")

In [37]:
geolocations_df.to_sql(
    name='geolocations', 
    con=db_engine, 
    if_exists='append', 
    index=False
)

logging.info("Loaded geolocations into database.")

In [38]:
customers_df.to_sql(
    name='customers', 
    con=db_engine, 
    if_exists='append', 
    index=False
)

logging.info("Loaded customers into database.")

In [39]:
sellers_df.to_sql(
    name='sellers', 
    con=db_engine, 
    if_exists='append', 
    index=False
)

logging.info("Loaded sellers into database.")

In [40]:
orders_df.to_sql(
    name='orders', 
    con=db_engine, 
    if_exists='append', 
    index=False
)

logging.info("Loaded orders into database.")

In [41]:
product_categories_df.to_sql(
    name='product_category_name_translations', 
    con=db_engine, 
    if_exists='append', 
    index=False
)

logging.info("Loaded product_category_name_translations into database.")

In [42]:
products_df.to_sql(
    name='products', 
    con=db_engine, 
    if_exists='append', 
    index=False
)

logging.info("Loaded products into database.")

In [43]:
order_reviews_df.to_sql(
    name='order_reviews', 
    con=db_engine, 
    if_exists='append', 
    index=False
)

logging.info("Loaded order_reviews into database.")

In [44]:
order_items_df.to_sql(
    name='order_items', 
    con=db_engine, 
    if_exists='append', 
    index=False
)

logging.info("Loaded order_items into database.")

In [45]:
order_payments_df.to_sql(
    name='order_payments', 
    con=db_engine, 
    if_exists='append', 
    index=False
)

logging.info("Loaded order_payments into database.")

In [46]:
# Close communication with server
cur.close()
conn.close()